src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1492,28 +1497,42 @@
           (select-keys migrations pending))))
 
 (defn migrate!
-  "Migrates database to the latest schema version. Does nothing if database is
-  already at the latest schema version."
-  []
+  "Migrates database to the latest schema version. Does nothing if
+  database is already at the latest schema version.  Requires a
+  connection pool because some operations may require an indepdendent
+  database connection."
+  [db-connection-pool]
   (if-let [unexpected (first (difference (applied-migrations) (kitchensink/keyset migrations)))]
     (throw (IllegalStateException.
             (format "Your PuppetDB database contains a schema migration numbered %d, but this version of PuppetDB does not recognize that version."
                     unexpected))))
-
-  (if-let [pending (seq (pending-migrations))]
-    (sql/transaction
-     (doseq [[version migration] pending]
-       (log/infof "Applying database migration version %d" version)
-       (try
-         (migration)
-         (record-migration! version)
-         (catch java.sql.SQLException e
-           (log/error e "Caught SQLException during migration")
-           (let [next (.getNextException e)]
-             (when-not (nil? next)
-               (log/error next "Unravelled exception")))
-           (System/exit 1)))))
-    (log/info "There are no pending migrations")))
+  (letfn [(sql-or-die [f]
+            (try
+              (f)
+              (catch java.sql.SQLException e
+                (log/error e "Caught SQLException during migration")
+                (let [next (.getNextException e)]
+                  (when-not (nil? next)
+                    (log/error next "Unravelled exception")))
+                ;; See PDB-1118
+                (System/exit 1))))]
+    (if-let [pending (seq (pending-migrations))]
+      (do
+        (sql/transaction
+         (doseq [[version migration] pending]
+           (log/infof "Applying database migration version %d" version)
+           (sql-or-die #(do (migration) (record-migration! version)))))
+        (when (sutils/postgres?)
+          ;; Make sure all tables (even small static tables) are
+          ;; analyzed at least once.  Note that vaccum cannot be
+          ;; called from within a transaction block.
+          (sql/with-connection db-connection-pool
+            (log/info "Analyzing database")
+            (sql-or-die #(-> (doto (:connection sqlint/*db*)
```

Comment:
With respect to the (fn [] ...), since it's short, I'd figured it was pretty easy to see there weren't any %'s in there, but I"m happy to change it if that's preferable.